packages/web_benchmarks/lib/src/timeseries.dart

```diff
@@ -0,0 +1,65 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+
+import 'metrics.dart';
+
+// TODO(kenz): move the time series logic from recorder.dart into this file for
+// better code separation.
+
+/// Computes the percentile threshold in [values] for the given [percentiles].
+///
+/// Each value in [percentiles] should be between 0.0 and 1.0.
+///
+/// Returns a [Map] of percentile values to the computed value from [values].
+Map<double, double> computePercentiles(
+  String label,
+  List<double> percentiles,
+  Iterable<double> values,
+) {
+  if (values.isEmpty) {
+    throw StateError(
+      '$label: attempted to compute a percentile of an empty value list.',
+    );
+  }
+  for (final double percentile in percentiles) {
+    if (percentile < 0.0 || percentile > 1.0) {
+      throw StateError(
+        '$label: attempted to compute a percentile for an invalid '
+        'value: $percentile',
+      );
+    }
+  }
+
+  final List<double> sorted =
+      values.sorted((double a, double b) => a.compareTo(b));
+  final Map<double, double> computed = <double, double>{};
+  for (final double percentile in percentiles) {
+    final int percentileIndex =
+        (sorted.length * percentile).round().clamp(0, sorted.length - 1);
+    computed[percentile] = sorted[percentileIndex];
+  }
+
+  return computed;
+}
+
+// ignore: public_member_api_docs, this will be made private in a follow up PR
+extension PercentileExtension on BenchmarkMetricComputation {
+  // ignore: public_member_api_docs, this will be made private in a follow up PR
+  double? get toPercentile {
+    if (RegExp(r'^p\d+$').hasMatch(name)) {
```

Comment:
Parsing enum names feels wrong somehow. Can we instead structure our enums in a way that doesn't require this? Here's one idea - use sealed classes with data fields instead of enums: https://gist.github.com/yjbanov/4639cffa4eec6db95a6fa0001c0cfe45

A simpler but less type-safe variant would be do attach `percentile` to the enum: https://gist.github.com/yjbanov/bfb52443c2f7db0121c55ecb4ed02f72

Reply:
used the sealed classes approach. Thanks.